pkg/controller/machinedeployment/rolling.go

```diff
@@ -0,0 +1,238 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machinedeployment
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/integer"
+
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	dutil "sigs.k8s.io/cluster-api/pkg/controller/machinedeployment/util"
+)
+
+// rolloutRolling implements the logic for rolling a new machine set.
+func (dc *MachineDeploymentControllerImpl) rolloutRolling(d *v1alpha1.MachineDeployment, msList []*v1alpha1.MachineSet, machineMap map[types.UID]*v1alpha1.MachineList) error {
+	newMS, oldMSs, err := dc.getAllMachineSetsAndSyncRevision(d, msList, machineMap, true)
+	if err != nil {
+		return err
+	}
+	allMSs := append(oldMSs, newMS)
+
+	// Scale up, if we can.
+	scaledUp, err := dc.reconcileNewMachineSet(allMSs, newMS, d)
+	if err != nil {
+		return err
+	}
+	if scaledUp {
+		return nil
+		// Update DeploymentStatus
+		//return dc.syncRolloutStatus(allMSs, newMS, d)
+	}
+
+	// Scale down, if we can.
+	scaledDown, err := dc.reconcileOldMachineSets(allMSs, dutil.FilterActiveMachineSets(oldMSs), newMS, d)
+	if err != nil {
+		return err
+	}
+	if scaledDown {
+		return nil
+		// Update DeploymentStatus
+		//return dc.syncRolloutStatus(allMSs, newMS, d)
+	}
+
+	if dutil.DeploymentComplete(d, &d.Status) {
+		if err := dc.cleanupDeployment(oldMSs, d); err != nil {
+			return err
+		}
+	}
+
+	return nil
+	// Sync deployment status
+	//return dc.syncRolloutStatus(allMSs, newMS, d)
+}
+
+func (dc *MachineDeploymentControllerImpl) reconcileNewMachineSet(allMSs []*v1alpha1.MachineSet, newMS *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) (bool, error) {
+	if *(newMS.Spec.Replicas) == *(deployment.Spec.Replicas) {
+		// Scaling not required.
+		return false, nil
+	}
+	if *(newMS.Spec.Replicas) > *(deployment.Spec.Replicas) {
+		// Scale down.
+		scaled, _, err := dc.scaleMachineSet(newMS, *(deployment.Spec.Replicas), deployment)
+		return scaled, err
+	}
+	newReplicasCount, err := dutil.NewMSNewReplicas(deployment, allMSs, newMS)
+	if err != nil {
+		return false, err
+	}
+	scaled, _, err := dc.scaleMachineSet(newMS, newReplicasCount, deployment)
+	return scaled, err
+}
+
+func (dc *MachineDeploymentControllerImpl) reconcileOldMachineSets(allMSs []*v1alpha1.MachineSet, oldMSs []*v1alpha1.MachineSet, newMS *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) (bool, error) {
+	oldMachinesCount := dutil.GetReplicaCountForMachineSets(oldMSs)
+	if oldMachinesCount == 0 {
+		// Can't scale down further
+		return false, nil
+	}
+
+	allMachinesCount := dutil.GetReplicaCountForMachineSets(allMSs)
+	glog.V(4).Infof("New machine set %s/%s has %d available machines.", newMS.Namespace, newMS.Name, newMS.Status.AvailableReplicas)
+	maxUnavailable := dutil.MaxUnavailable(*deployment)
+
+	// Check if we can scale down. We can scale down in the following 2 cases:
```

Comment:
This is the same logic as the one from deployment.
Would the max unavailable setting help prevent the worst case scenario, where some number of machines needs to be healthy before proceeding?

Reply:
Given this is the deployment model and that seems to work in practice, I think that trumps my intuition :-)  I'm sure we'll learn new things when we do this on real machines!

Reply:
We will definitely learn more once we get people using it.
After we get more usage, and deem that we want to stay in sync with the deployment controller logic, I will be reaching out to the kubernetes team to talk about possibilities of sharing their logic to avoid the duplication here.